src/cms.c

```diff
@@ -31,66 +43,84 @@ void CMS_Destroy(CMSketch *cms) {
     CMS_FREE(cms);
 }
 
-void CMS_IncrBy(CMSketch *cms, const char *item, size_t value) {
+void CMS_IncrBy(CMSketch *cms, const char *item, size_t strlen, size_t value) {
     assert(cms);
     assert(item);
 
     for(size_t i = 0; i < cms->depth; ++i) {
-        size_t hash = XXH64(item, strlen(item), i);
+        uint32_t hash = MurmurHash2(item, strlen, i);
         cms->array[(hash % cms->width) + (i * cms->width)] += value;
     }
     cms->counter += value;
 }
 
-size_t CMS_Query(CMSketch *cms, const char *item) {
+size_t CMS_Query(CMSketch *cms, const char *item, size_t strlen) {
     assert(cms);
     assert(item >= 0);
 
-    size_t temp = 0, res = ~0;
+    size_t temp = 0, res = -1;
 
     for(size_t i = 0; i < cms->depth; ++i) {
-        size_t hash = XXH64(item, strlen(item), i);
+        uint hash = MurmurHash2(item, strlen, i);
         temp = cms->array[(hash % cms->width) + (i * cms->width)];
-        if(temp < res) res = temp;
+        if(temp < res) {
+            res = temp;
+        }
     }
 
     return res;
 }
 
-void CMS_Merge(CMSketch *dest, size_t quantity, CMSketch **src, long long *weight) {
+void CMS_Merge(CMSketch *dest, size_t quantity, 
+               const CMSketch **src, const long long *weights) {
     assert(dest);
     assert(src);
-    assert(weight);
+    assert(weights);
 
-    size_t tempCount= 0;
-    size_t tempTotal = 0;
+    size_t itemCount= 0;
+    size_t cmsCount = 0;
     size_t width = dest->width;
     size_t depth = dest->depth;
     
     for(size_t i = 0; i < depth; ++i) {
         for(size_t j = 0; j < width; ++j) {            
-            tempCount = 0;
+            itemCount = 0;
             for(size_t k = 0; k < quantity; ++k) {
-                tempCount += src[k]->array[(i * width) + j] * weight[k];
+                itemCount += src[k]->array[(i * width) + j] * weights[k];
             }
-            dest->array[(i * width) + j] = tempCount; 
+            dest->array[(i * width) + j] = itemCount; 
         }
     }
 
     for(size_t i = 0; i < quantity; ++i) {
-        tempTotal += src[i]->counter * weight[i];
+        cmsCount += src[i]->counter * weights[i];
     }
-    dest->counter = tempTotal;
+    dest->counter = cmsCount;
 }
  
 void CMS_Print(const CMSketch *cms) {
     assert(cms);
 
     for(int i = 0; i < cms->depth; ++i) {
         for(int j = 0; j < cms->width; ++j) {
-            printf("%lu\t", cms->array[(i * cms->width) + j]);
+            printf("%d\t", cms->array[(i * cms->width) + j]);
         }
         printf("\n");
     }
-    printf("\tCounter is %lu\n", (size_t)cms->counter);
+    printf("\tCounter is %lu\n", cms->counter);
+}
+
+size_t CMS_cardinality(CMSketch *cms) {
```

Comment:
CMS_GetCardinality

Reply:
It isn't yet integrated in.
For now, I'll probably add this to cms.info